File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
private:
    bool solve(vector<int>& nums, int k){
        int n = nums.size();
        vector<vector<int>> dp(n, vector<int>(k+1));
        
        for(int i=0;i<n;i++)dp[i][0] = true;
        // dp[0][nums[0]] = true;

        for(int idx =1;idx<n;idx++){
            for(int tar=1;tar<=k;tar++){
                bool  nt = dp[idx-1][tar];
                bool t = false;
                if(nums[idx]<=tar)
                t = dp[idx-1][tar-nums[idx]];

                dp[idx][tar] = t | nt;
            }
        }
        return dp[n-1][k];
    }
public:
    bool canPartition(vector<int>& nums) {
        int sum=0;
        for(auto i : nums)
        sum += i;

        if(sum%2!=0)return false;
        return solve(nums, sum/2);
    }
};

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
    public int solve(int n, int k, int arr[]){
        int dp[][] = new int[n+1][k+1];

        for(int i =0;i<=n;i++)dp[i][0] =1;

        for(int i =1;i<=n;i++){
            for(int j=0;j<=k;j++){
                int np = dp[i-1][j];
                int p =0;
                if(j >= arr[i-1])
                p = dp[i-1][j-arr[i-1]];

                dp[i][j] = p | np;
            }
        }
        return dp[n][k];
    }
    public boolean canPartition(int[] nums) {
        int sum = Arrays.stream(nums).sum();

        if(sum%2 !=0)return false;
        return solve(nums.length, sum/2 ,nums)==1;
    }
}